src/array.jl

```diff
@@ -738,3 +738,82 @@ summary(A::CategoricalArray{T, N, R}) where {T, N, R} =
 
 refs(A::CategoricalArray) = A.refs
 pool(A::CategoricalArray) = A.pool
+
+"""
+    sort!(V::CategoricalVector; rev::Bool=false)
+
+A fast in-place sorting method for `CategoricalVector`` using counting sort
+""" 
+function is_categorical_after_by(by::Function, v::CategoricalVector)
+    for v1 in v
+        if !ismissing(v1) 
+            return isa(by(v1), CatValue) 
+        elseif isa(by(v1), CatValue) # if v1 is missing but `by`` still converts it to a `CatValue`
+            return true
+        end
+    end
+    # if all values were missing then it will reach here
+    return false
+end
+
+function Base.sort!(v::CategoricalVector;
+                    alg::Base.Algorithm=Base.Sort.defalg(v),
+                    lt=isless,
+                    by=identity,
+                    rev::Bool=false,
+                    order::Base.Ordering=Base.Forward)
+
+    ordr = Base.ord(lt,by,rev,order)
+    if (ordr == Base.Forward || ordr == Base.Reverse) && is_categorical_after_by(by, v)
+        if ordr == Base.Forward
+            rev = false
+        else ordr == Base.Reverse
```

Comment:
This syntax is incorrect. You can just do `rev = (order == Base.Reverse)`.

Reply:
done